### ISSUE TITLE
Issues with cyrillic (utf-8) when training 

### ISSUE THREAD

**Issue by alekol** [RaRe-Technologies/gensim#1326]
#### Description
I'm training word2vec with Bulgarian Wiki corpora. I am using python 3.5 and during training I'm getting errors in word2vec.update_weights() and word2vec.reset_weights() because implicit concatenation of byte array and strings isn't possible:

```python
def update_weights(self):
        ...
        # randomize the remaining words
        for i in xrange(len(self.wv.syn0), len(self.wv.vocab)):
            # construct deterministic seed from word AND seed argument
            newsyn0[i-len(self.wv.syn0)] = self.seeded_vector(self.wv.index2word[i] + str(self.seed))
```
```python
def reset_weights(self):
        ...
        for i in xrange(len(self.wv.vocab)):
            # construct deterministic seed from word AND seed argument
            self.wv.syn0[i] = self.seeded_vector(self.wv.index2word[i] + str(self.seed))
```

The exceptions are thrown on the last line of each of the samples above

#### Steps/Code/Corpus to Reproduce
```python
import os
import multiprocessing
from gensim.corpora.wikicorpus import WikiCorpus
from gensim.models.word2vec import Word2Vec
from gensim.corpora import Dictionary

import logging

dictionary_name = 'bgwiki_wordids.txt.bz2'
dump_name = 'bgwiki-latest-pages-articles.xml.bz2'

if __name__ == '__main__':

    logger = logging.getLogger('w2v_logger')

    logging.basicConfig(format='%(asctime)s: %(levelname)s: %(message)s')
    logging.root.setLevel(level=logging.INFO)

    logger.info("building dictionary")
    if not os.path.isfile(dictionary_name):
        logging.info('Dictionary has not been created yet..')
        logging.info('Creating dictionary (takes about 9h)..')

        # Construct corpus
        wiki = WikiCorpus(dump_name)

        # Remove words occuring less than 20 times, and words occuring in more
        # than 10% of the documents. (keep_n is the vocabulary size)
        wiki.dictionary.filter_extremes(no_below=10, no_above=0.1, keep_n=100000)

        # Save dictionary to file
        wiki.dictionary.save_as_text(dictionary_name)
        del wiki

    logger.info("loading dictionary")
    dictionary = Dictionary.load_from_text(dictionary_name)

    logger.info("loading corpus")
    wiki = WikiCorpus(dump_name, dictionary=dictionary)

    logger.info("getting sentences")
    sentences = list(wiki.get_texts())
    params = {'size': 400, 'window': 10, 'min_count': 10,
              'workers': max(1, multiprocessing.cpu_count() - 1), 'sample': 1E-3}

    logger.info("training the model")
    word2vec = Word2Vec(sentences, **params)

    logger.info("storing the model")
    word2vec.save('bgwiki_word2vec')
```
The Bulgarian Wiki corpora can be downloaded here: https://dumps.wikimedia.org/bgwiki/latest/bgwiki-latest-pages-articles.xml.bz2

#### Expected Results
Model is trained properly

#### Actual Results
Model training fails. Fixing the concatenation with gensim.utils.to_unicode() helps with training, but then exploring the model fails (indexing with u'мъж' doesn't work).

More on that last problem here: https://groups.google.com/forum/#!topic/gensim/KWfl_wdU37E

#### Versions
Windows-10-10.0.14393-SP0
Python 3.5.2 |Continuum Analytics, Inc.| (default, Jul  5 2016, 11:41:13) [MSC v.1900 64 bit (AMD64)]
NumPy 1.12.1
SciPy 0.19.0
gensim 2.1.0
FAST_VERSION 1

**Comment by tmylk** [comment]
Closing as it's being discussed on the mailing list.

**Comment by alekol** [comment]
Well, there are 3 distinct py3 and unicode-related bugs that I have found and fixed locally so far.

Could you please reopen the issue? I'd like to contribute back the fixes for these

**Comment by tmylk** [comment]
Looking forward to the fixes

**Comment by menshikh-iv** [comment]
Partial resolved in #1333
